# Problem #1

## (a) Hello World

The "Hello World" program is simple in Python 3:

In [1]:
print('Hello, World')

Hello, World


## (b) Printing practice

In [2]:
print('%.1e\n' % 19000000000)

1.9e+10



In [2]:
import numpy as np
print ('The value of the golden mean is %.8f' % ((np.sqrt(5)-1)/2))

The value of the golden mean is 0.61803399


Comments:     
`%e` inserts a "Floating point exponential format (lowercase)" value  
`%.1e` tells python to keep only one decimal place  
`\n` is the newline character, and inserts a blank line after the print statement. This would make the output neater if we were printing multiple lines. 

# Problem 2

For this problem, we use `float64` for "double precision" and `float32` for "single precision."

## (a) Calculation in double precision

We pick N=1000 rather arbitrarily.  Be careful with the ranges here, making sure to get n=1 and n=1000 in both sums.

In [3]:
N = 1000
sum_up64 = np.float64(0.)
sum_down64 = np.float64(0.)

# We need to include term N, so extend the range by 1
for n in range (1,N+1,1):
    sum_up64 += np.float64(1./n)

for n in range (N,0,-1):
    sum_down64 += np.float64(1./n)

print("64-bit S^{up} =   %.16f" % sum_up64)
print("64-bit S^{down} = %.16f" % sum_down64)

64-bit S^{up} =   7.4854708605503433
64-bit S^{down} = 7.4854708605503406


These seem to be the same, up to the machine precision (or close to it).

## (b) Comparison of up and down

In [4]:
print("64-bit difference [up-down] is %e" % (sum_up64 - sum_down64))

64-bit difference [up-down] is 2.664535e-15


This is a more formal test than the quick check by eye in part (a).  The 64-bit result should count as "high accuracy," because it is near the limit of 64-bit floating-point precision.

## (c) Calculation with single precision

Now we calculate the sums for different values of N.

In [5]:
for p in (2,3,4,5,6,7):
    N = pow(10,p)
    
    sum_up32 = np.float32(0.)
    sum_down32 = np.float32(0.)

    for n in range (1,N+1,1):
        sum_up32 += np.float32(1./n)

    for n in range (N,0,-1):
        sum_down32 += np.float32(1./n)

    print("For N = %i" % N)
    print("32-bit S^{up} =   %.16f" % sum_up32)
    print("32-bit S^{down} = %.16f" % sum_down32)   

For N = 100
32-bit S^{up} =   5.1873779296875000
32-bit S^{down} = 5.1873769760131836
For N = 1000
32-bit S^{up} =   7.4854784011840820
32-bit S^{down} = 7.4854717254638672
For N = 10000
32-bit S^{up} =   9.7876129150390625
32-bit S^{down} = 9.7876043319702148
For N = 100000
32-bit S^{up} =   12.0908508300781250
32-bit S^{down} = 12.0901527404785156
For N = 1000000
32-bit S^{up} =   14.3573579788208008
32-bit S^{down} = 14.3926515579223633
For N = 10000000
32-bit S^{up} =   15.4036827087402344
32-bit S^{down} = 16.6860313415527344


## (d) Error scaling

Repeat part (c), but this time including the `double` (`float64`) precision and comparing to the `single` precision.

In [6]:
for p in (2,3,4,5,6,7):
    N = pow(10,p)
    
    sum_up32 = np.float32(0.)
    sum_down32 = np.float32(0.)
    sum_up64 = np.float64(0.)
    sum_down64 = np.float64(0.)

    for n in range (1,N+1,1):
        sum_up32 += np.float32(1./n)
        sum_up64 += np.float64(1./n)

    for n in range (N,0,-1):
        sum_down32 += np.float32(1./n)
        sum_down64 += np.float64(1./n)

    print("For N = %i" % N)
    print("64-bit S^{up} - 32-bit S^{up} =     %19.16f" % (sum_up64-sum_up32))
    print("64-bit S^{down} - 32-bit S^{down} = %19.16f" % (sum_down64-sum_down32))   
    print()

    

For N = 100
64-bit S^{up} - 32-bit S^{up} =     -0.0000004120478794
64-bit S^{down} - 32-bit S^{down} =  0.0000005416264379

For N = 1000
64-bit S^{up} - 32-bit S^{up} =     -0.0000075406337388
64-bit S^{down} - 32-bit S^{down} = -0.0000008649135266

For N = 10000
64-bit S^{up} - 32-bit S^{up} =     -0.0000068789947143
64-bit S^{down} - 32-bit S^{down} =  0.0000017040741707

For N = 100000
64-bit S^{up} - 32-bit S^{up} =     -0.0007047002147900
64-bit S^{down} - 32-bit S^{down} = -0.0000066106151078

For N = 1000000
64-bit S^{up} - 32-bit S^{up} =      0.0353687440441881
64-bit S^{down} - 32-bit S^{down} =  0.0000751649434090

For N = 10000000
64-bit S^{up} - 32-bit S^{up} =      1.2916286571170374
64-bit S^{down} - 32-bit S^{down} =  0.0092800243072304



We can see that the error on $S^\text{(up)}$ is always greater than the error on $S^\text{(down)}$, except for N=100.  

We also see that the errors on both $S^\text{(up)}$ and $S^\text{(down)}$ increase with N, but the error on $S^\text{(up)}$ (and therefore the difference) increases faster.

## (e) Explanation of error scaling

The reason for this effect is the limited precision of the floating-point representation.

When we add relatively large numbers (1, 0.5, 0.25), followed by small numbers, we reach a point where the small numbers are never contributing anything because their values are less than the machine precision.


In [7]:
1.000000000000000 + 0.0000000000000001

1.0

If instead we add all of the small numbers together first (large N), then we get an accurate sum that is not small and can be added accurately to the big numbers.

The moral of this exercise is: *don’t add a lot of small numbers to a big one.*

# Problem 3

## (a) nth power of $\phi$

The successive multiplication just means $\phi^4 = \phi \cdot \phi^3$ and so on.

We'll do the first 50 powers, to match part (c).

In [8]:
phi = (np.sqrt(5)-1)/2
powers_of_phi = []
powers_of_phi.append(1)
for n in range (1,51):
    powers_of_phi.append(phi * powers_of_phi[n-1])
    print(n, powers_of_phi[n])

1 0.6180339887498949
2 0.3819660112501052
3 0.23606797749978975
4 0.14589803375031551
5 0.09016994374947429
6 0.05572809000084125
7 0.03444185374863305
8 0.021286236252208206
9 0.01315561749642485
10 0.008130618755783357
11 0.005024998740641495
12 0.003105620015141862
13 0.0019193787254996341
14 0.0011862412896422286
15 0.000733137435857406
16 0.00045310385378482284
17 0.00028003358207258323
18 0.00017307027171223967
19 0.00010696331036034358
20 6.61069613518961e-05
21 4.085634900844749e-05
22 2.5250612343448615e-05
23 1.560573666499888e-05
24 9.64487567844974e-06
25 5.960860986549141e-06
26 3.6840146919005996e-06
27 2.2768462946485426e-06
28 1.4071683972520572e-06
29 8.696778973964855e-07
30 5.374904998555718e-07
31 3.321873975409138e-07
32 2.05303102314658e-07
33 1.2688429522625587e-07
34 7.841880708840217e-08
35 4.846548813785372e-08
36 2.995331895054845e-08
37 1.8512169187305277e-08
38 1.144114976324318e-08
39 7.071019424062098e-09
40 4.370130339181083e-09
41 2.7008890848810156e-09

## (b) Recursion relation derivation

The easiest way to show the recursion relation holds for $\phi$ is to multiply the given relation by $\phi^{-(n-1)}$:

$$
\begin{align}
\phi^{n+1} &= \phi^{n-1} - \phi^n \\
\phi^2 &= \phi^0 - \phi^1 \\
\phi^2 + \phi -1 &= 0 
\end{align}
$$

This is a quadratic equation, whose roots are 
$$\phi = (-1 \pm \sqrt{5})/2$$

The positive solution gives the golden mean $\phi = (\sqrt{5}-1)/2 \simeq 0.618\dots$.

This proves the relation.

## (c) Implementation of subtractive recursion



In [9]:
phi = (np.sqrt(5)-1)/2
recursive_powers_of_phi32 = []
recursive_powers_of_phi32.append(np.float32(1))
recursive_powers_of_phi32.append(np.float32(phi))
recursive_powers_of_phi64 = []
recursive_powers_of_phi64.append(np.float64(1))
recursive_powers_of_phi64.append(np.float64(phi))
print("n \t  exact \t  single \t  double")
for n in range (1,50):
    # We are actually appending the (n+1)th power of phi
    recursive_powers_of_phi32.append(recursive_powers_of_phi32[n-1] 
                                   - recursive_powers_of_phi32[n])
    recursive_powers_of_phi64.append(recursive_powers_of_phi64[n-1] 
                                   - recursive_powers_of_phi64[n])
    print("%d \t %13.6e \t %13.6e \t %13.6e" % (n+1, 
                                    powers_of_phi[n+1], 
                                    recursive_powers_of_phi32[n+1], 
                                    recursive_powers_of_phi64[n+1]))


n 	  exact 	  single 	  double
2 	  3.819660e-01 	  3.819660e-01 	  3.819660e-01
3 	  2.360680e-01 	  2.360680e-01 	  2.360680e-01
4 	  1.458980e-01 	  1.458980e-01 	  1.458980e-01
5 	  9.016994e-02 	  9.017003e-02 	  9.016994e-02
6 	  5.572809e-02 	  5.572796e-02 	  5.572809e-02
7 	  3.444185e-02 	  3.444207e-02 	  3.444185e-02
8 	  2.128624e-02 	  2.128589e-02 	  2.128624e-02
9 	  1.315562e-02 	  1.315618e-02 	  1.315562e-02
10 	  8.130619e-03 	  8.129716e-03 	  8.130619e-03
11 	  5.024999e-03 	  5.026460e-03 	  5.024999e-03
12 	  3.105620e-03 	  3.103256e-03 	  3.105620e-03
13 	  1.919379e-03 	  1.923203e-03 	  1.919379e-03
14 	  1.186241e-03 	  1.180053e-03 	  1.186241e-03
15 	  7.331374e-04 	  7.431507e-04 	  7.331374e-04
16 	  4.531039e-04 	  4.369020e-04 	  4.531039e-04
17 	  2.800336e-04 	  3.062487e-04 	  2.800336e-04
18 	  1.730703e-04 	  1.306534e-04 	  1.730703e-04
19 	  1.069633e-04 	  1.755953e-04 	  1.069633e-04
20 	  6.610696e-05 	 -4.494190e-05 	  6.610696e-05
21 	  4.

The recursion relation is not stable for large $n$.

We see that the single precision recursion matches the exact value well until $n\simeq 14$, after which it starts to diverge.

The double precision recursion matches the exact value until $n\simeq 33$.

## (d) General solution for recursion relation

Recall that there were actually two roots to the quadratic form of the recursion relation.  The first root was the golden mean $\phi = (-1 + \sqrt{5})/2 \simeq 0.618 \dots$, and the second root was $\tilde{\phi} = (-1 - \sqrt{5})/2 \simeq -1.618 \dots$.

Since both of these satisfy the quadratic relation, the general solution to the recursion relation is 
$$\Phi[n] = A\phi[n] + B\tilde{\phi}[n],$$
where $A$ and $B$ are arbitrary constants.

In principle, we can specify the use of one or the other solution by setting, for example, $A=1$ and $B=0$.  This is actually the boundary condition we set in parts (b) and (c).

The problem is that $A=1$ is not exact in machine representation, and a little bit of $\tilde{\phi}$ starts to sneak into our solution, and we end up with a function

$$\Phi[n] = \phi[n] + \epsilon \tilde{\phi},$$
where $\epsilon$ is the machine precision.

As we exponentiate $\Phi$, we pick up more and more of $\tilde{\phi}$, simply because $|\tilde{\phi}| > |\phi|$.  Eventually the algorithm becomes unstable.

# Problem 4

## (a) Analytic derivation of midpoint expression and error

We can derive this expression by using the Taylor expansion of the function at some specific point $x$.

$$f(x+h) = f(x) + h \frac{df}{dx} + \frac{h^2}{2!}\frac{d^2 f}{dx^2} + \frac{h^3}{3!} \frac{d^3 y}{dx^3} + \frac{h^4}{4!}\frac{d^4 f}{dx^4} + \cdots$$

Therefore,
$$
\begin{align}
\frac{f(x+h)-f(x-h)-2f(x)}{h^2} &= \frac{\left[ f(x) + h \frac{df}{dx} + \frac{h^2}{2!}\frac{d^2 f}{dx^2} + \frac{h^3}{3!} \frac{d^3 y}{dx^3} + \frac{h^4}{4!}\frac{d^4 f}{dx^4} + \cdots \right] + \left[ f(x) - h \frac{df}{dx} + \frac{h^2}{2!}\frac{d^2 f}{dx^2} - \frac{h^3}{3!} \frac{d^3 y}{dx^3} + \frac{h^4}{4!}\frac{d^4 f}{dx^4} + \cdots \right] - 2f(x)}{h^2}\\
&= \frac{h^2 \frac{d^2 f}{dx^2} + \frac{h^4}{12}\frac{d^4 f}{dx^4} + \cdots}{h^2} \\
&= f^{\prime \prime}(x) + \frac{h^2}{12} f^{(4)}(x) \\
\end{align}
$$

Therefore,
$$f^{\prime \prime}(x) = \frac{f(x+h)-f(x-h)-2f(x)}{h^2} - \frac{h^2}{12} f^{(4)}(x),$$
where the second term on the right-hand side can be considered as the error.

## (b) Numerical testing

We will use the function $f(x) = \sin(x)$ and evaluate its derivative $f^{\prime\prime}(x)$ at $x=\pi/4$.  This numerical derivative is compared to the exact value of $-\sin(x)$.

In [10]:
x = np.pi/4
print("h \t \t f''_midpoint \t d2f/dx2 \t error")
for p in range(7):
    h = pow(10,-p)
    f2_midpoint = (np.sin(x+h) + np.sin(x-h) - 2*np.sin(x))/(h*h)
    print("%e \t %f \t %f \t %.12f" %(h, f2_midpoint, -np.sin(x), 
                             -np.sin(x)-f2_midpoint))

h 	 	 f''_midpoint 	 d2f/dx2 	 error
1.000000e+00 	 -0.650111 	 -0.707107 	 -0.056996067554
1.000000e-01 	 -0.706518 	 -0.707107 	 -0.000589059268
1.000000e-02 	 -0.707101 	 -0.707107 	 -0.000005892538
1.000000e-03 	 -0.707107 	 -0.707107 	 -0.000000058900
1.000000e-04 	 -0.707107 	 -0.707107 	 -0.000000008052
1.000000e-05 	 -0.707105 	 -0.707107 	 -0.000001295911
1.000000e-06 	 -0.706990 	 -0.707107 	 -0.000116759105


We notice that the error decreases with $h^2$, as expected, until $h<10^{-4}$, when round-off errors start to creep in.

# Problem 5

We will compute the spherical Bessel functions by using the downward recurrence relation

$$j_{l-1}(x) = \frac{2l+1}{x} j_l(x) - j_{l+1}(x)$$

This is not enough.  We need to scale the values so that $j_0(x)$ matches the exact value $$j_0(x) = \sin(x)/x$$

The results are given for the first 25 $l$ values at $x=0.1, 1.0, 10.$

In [11]:
from numpy import *
for x in (0.1, 1., 10.):
    print("Working at x =", x)
    j = zeros(52)
    j[51] = 1.0
    j[50] = 1.0
    for l in range(50, 0, -1):
    #     print("l =", l, ", j[l] =", j[l])
        j[l-1] = (2*l+1)/x * j[l] - j[l+1]
    scale_factor = j[0] * x / sin(x)
    for l in range(25):
        print("l =", l, ", j[l] =", j[l]/scale_factor)

Working at x = 0.1
l = 0 , j[l] = 0.9983341664682815
l = 1 , j[l] = 0.03330001190255757
l = 2 , j[l] = 0.0006661906084455687
l = 3 , j[l] = 9.518519720865566e-06
l = 4 , j[l] = 1.057720150209873e-07
l = 5 , j[l] = 9.616310232916444e-10
l = 6 , j[l] = 7.397541093587706e-12
l = 7 , j[l] = 4.9318874757319734e-14
l = 8 , j[l] = 2.9012001025301897e-16
l = 9 , j[l] = 1.5269856934948196e-18
l = 10 , j[l] = 7.271510996713672e-21
l = 11 , j[l] = 3.1615815051510697e-23
l = 12 , j[l] = 1.264651337875089e-25
l = 13 , j[l] = 4.6839536652559885e-28
l = 14 , j[l] = 1.6151744028156545e-30
l = 15 , j[l] = 5.210290941008979e-33
l = 16 , j[l] = 1.5788897128763593e-35
l = 17 , j[l] = 4.5111483007244514e-38
l = 18 , j[l] = 1.2192377198447579e-40
l = 19 , j[l] = 3.126270115223257e-43
l = 20 , j[l] = 7.625092312409071e-46
l = 21 , j[l] = 1.7732864462699687e-48
l = 22 , j[l] = 3.9406551792868825e-51
l = 23 , j[l] = 8.384409128498707e-54
l = 24 , j[l] = 1.711110750982455e-56
Working at x = 1.0
l = 0 , j[l] = 0

Now we will try the upward recursion:

$$j_{l+1}(x) = \frac{2l+1}{x} j_l(x) - j_{l-1}(x)$$

We start with known values of $j[0]$ and $j[1]$.

In [12]:
from numpy import *
for x in (0.1, 1., 10.):
    print("Working at x =", x)
    j = zeros(52)
    j[0] = sin(x)/x
    j[1] = (sin(x)-x*cos(x))/x/x
    for l in range(2,50):
        j[l+1] = (2*l+1)/x * j[l] - j[l-1]
    for l in range(25):
        print("l =", l, ", j[l] =", j[l])

Working at x = 0.1
l = 0 , j[l] = 0.9983341664682815
l = 1 , j[l] = 0.0333000119025581
l = 2 , j[l] = 0.0
l = 3 , j[l] = -0.0333000119025581
l = 4 , j[l] = -2.331000833179067
l = 5 , j[l] = -209.75677497421347
l = 6 , j[l] = -23070.914246330303
l = 7 , j[l] = -2999009.0952479653
l = 8 , j[l] = -449828293.37294847
l = 9 , j[l] = -76467810864.306
l = 10 , j[l] = -14528434235924.768
l = 11 , j[l] = -3050894721733336.5
l = 12 , j[l] = -7.016912575644315e+17
l = 13 , j[l] = -1.7541976349638613e+20
l = 14 , j[l] = -4.73626344527667e+22
l = 15 , j[l] = -1.3734988571538846e+25
l = 16 , j[l] = -4.257799094542589e+27
l = 17 , j[l] = -1.4050599662104828e+30
l = 18 , j[l] = -4.917667303745745e+32
l = 19 , j[l] = -1.8195228517862633e+35
l = 20 , j[l] = -7.096089945293389e+37
l = 21 , j[l] = -2.909378682341772e+40
l = 22 , j[l] = -1.2510257373170167e+43
l = 23 , j[l] = -5.629586724139752e+45
l = 24 , j[l] = -2.64589325008831e+48
Working at x = 1.0
l = 0 , j[l] = 0.8414709848078965
l = 1 , j[l] = 0.3

We conclude that the upward recursion relation is unstable and cannot give reliable results for large values of $l$.